### DISCUSSION TITLE
Return type is only added to an interface but implementations are skipped

### DISCUSSION THREAD

**Discussion by mxr576** [rectorphp/rector#7267] [Q&A] [answered]
TL;DR Can missing (union) return typehint be added to Foo class not just FooContract by Rector automatically?

I was playing around with Rector to try to update our (Drupal) codebase from PHP 7.4 and 8.1 and based on what I saw, Rector could only add union return type to an interface but not to those classes that implements the interface which lead to fatal error. What am I missing? (It looks like having `@inheritDoc` or `{@inheritdoc}` on `Foo::do()` does not change anything.)

```
<?php

interface FooContract {

  /**
   * @return Bar|Baz
   */
  public function do();
}

class Foo implements FooContract {

  public function do() {
    return new Bar();
  }

}

final class Bar {

}

final class Baz {

}
```

**Comment by mxr576** [comment]
@TomasVotruba Hi, sorry for the direct ping but can you share some insights about how/if this should work?

**Reply by TomasVotruba** [reply]
Hi, no troubles. I would like to help, but I don't understand the quesiton. Are you making custom rule or trying to use some existing one? I miss the `rector.php` configuration for latter one.

The demo link with minimal example might help: https://getrector.org/demo

**Reply by mxr576** [reply]
Sorry, I was not clear about that. I try to leverage Rector's built-in rules to upgrade a PHP 7.4 code to PHP 8.1. Based on my experience Rector 0.13.7 can add native union return types to `FooContract::do()` but leaves out `Foo::do()` which leads to a fatal error after Rector is done.

My question: Is it a known limitation of Rector or am I miss configuring something? How can both the interface and the class refactored and return types added to both methods?

I also played on Demo, but I could not even achieve what I could in my local, that `FooContract::do()` was refactored. This is one of my trials. not the best but I do not have the complete history now:  https://getrector.org/demo/a440e9e0-ac00-41bc-b04c-60840bae3657


My expected outcome would be:

```php
<?php

interface FooContract {

  /**
   * @return Bar|Baz
   */
  public function do(): Bar|Baz;
}

class Foo implements FooContract {

  public function do() : Bar|Baz {
    // @fixme
    return new Bar();
  }

}

final class Bar {

}

final class Baz {

}

```

**Reply by TomasVotruba** [reply]
I see. Seems the demo link does not modify any code. Is it correct?

The typed upgrade have to be sure the return type is always valid, so there might be some condition that stops it.
You can check existing fixtures for `UnionTypesRector` to find similar example:
https://github.com/rectorphp/rector-src/tree/main/rules-tests/Php80/Rector/FunctionLike/UnionTypesRector/Fixture

Yet the more helpful rules you're looking for might be in TYPE_DECLARATION/_STRICT sets.

**Reply by mxr576** [reply]
Thanks for the tips, with the help of TYPE_ rules, Rector also updated return type on classes, not just on interfaces.
My config is:

```php
  $rectorConfig->sets([
    LevelSetList::UP_TO_PHP_81,
    SetList::TYPE_DECLARATION,
    SetList::TYPE_DECLARATION_STRICT, // Not strictly necessary for the expected behavior (pun intended).
  ]);

```

Although I am still not entirely sure what is going on:
1. Is it an expected behavior that LevelSetList::UP_TO_PHP_81 is not enough for updating a PHP 7.4 codebase to PHP 8.1?
2. Why the above configuration still does nothing on Demo? https://getrector.org/demo/a8850128-2236-408b-94d4-a1cc14041638 (This is the reason why I shared screenshots too.)
3. (Could be unrelated/false positive) Some implementation of the interface (on the screenshot) still did not get the proper union type... needs investigation.


![image](https://user-images.githubusercontent.com/1755573/177290422-0ee92b00-28d5-44e9-be6b-2d895d5f21d8.png)

![image](https://user-images.githubusercontent.com/1755573/177289106-478d31d8-911e-4f49-bd21-97b5321b13df.png)

**Reply by samsonasik** [reply]
You need to return new Bar(); instead of Bar();:

```diff
-return Bar();
+return new Bar();
```

ref https://getrector.org/demo/c6b942b4-2cdb-4ac7-b5e5-e8f3c77a62ad

**Reply by mxr576** [reply]
Aaaaahhhhh.... :facepalm: 

Does it also mean that when a class method's return type is refactored, the "inherited (union) return type" is not considered, rather only the possible return values that the given implementation can produce? :thinking: 

https://getrector.org/demo/e820f7df-429f-46eb-9532-32a241802a9e

**Reply by samsonasik** [reply]
You may looking for `AddReturnTypeDeclarationRector` for that https://getrector.org/demo/c7564af9-c139-4cab-9f94-95ab1716c6a1

**Reply by mxr576** [reply]
That is also a viable option indeed. Thx!

**Comment by mxr576** [comment, marked as answer]
So for the record the actual solution is something like: 
https://getrector.org/demo/923fc431-82bd-424a-b1a5-1e72bdaf161b
or
https://getrector.org/demo/e820f7df-429f-46eb-9532-32a241802a9e
or 
https://getrector.org/demo/c7564af9-c139-4cab-9f94-95ab1716c6a1
